File: 207_Course Schedule.py
class Solution:
    # # 1. Topological Sort
    # def canFinish(self, numCourses: int, preq: List[List[int]]) -> bool:
    #     cnext = [[] for _ in range(numCourses)]
    #     inDeg = [0] * numCourses
        
    #     # 1. init
    #     for a, b in preq:
    #         cnext[b].append(a)
    #         inDeg[a] += 1
        
    #     # 2. loop to process those w/n in-degrees
    #     q = deque()
    #     for i, d in enumerate(inDeg):
    #         if d == 0: q.append(i)
        
    #     cnt = 0
    #     while q:
    #         tmp = q.pop()
    #         cnt += 1

    #         for cn in cnext[tmp]:
    #             inDeg[cn] -= 1
    #             if inDeg[cn] == 0: q.append(cn)
    #     return cnt == numCourses

    # 2. DFS
    def canFinish(self, numCourses: int, preq: List[List[int]]) -> bool:
        # 1. init w/ prep
        cprev = [[] for _ in range(numCourses)]
        for a, b in preq: cprev[a].append(b)

        # 2. init the valid status (0-not check, 1-checking, 2-valid)
        valid = [0] * numCourses

        # 3. dfs for each course, check if each can be completed
        def bt(course: int) -> bool:
            # end case
            if valid[course] == 2: return True
            if valid[course] == 1: return False
            # process
            valid[course] = 1

            for cp in cprev[course]:
                if not bt(cp): return False
            
            valid[course] = 2
            return True
        
        for c in range(numCourses):
            if not bt(c): return False
        return True

File: 207_Course Schedule.java
// Time complexity: O(V+E)
// Space complexity: O(V+E)
// V is the number of courses and E is the number of prerequisites.

class Solution {
    // // Method 1. Topological Sort
    // public boolean canFinish(int numCourses, int[][] preq) {
    //     // 1. build adj-list & record in-degree
    //     Map<Integer, List<Integer>> cnext = new HashMap<>();
    //     int[] inDegree = new int[numCourses];

    //     for (int i = 0; i < numCourses; i++) cnext.put(i, new ArrayList<>());
    //     for (int[] i : preq) {
    //         cnext.get(i[1]).add(i[0]);
    //         inDegree[i[0]] += 1;
    //     }
        
    //     // 2. start from those w/n in-degree
    //     int cnt = 0;
    //     Deque<Integer> q = new LinkedList<>();
    //     for (int i = 0; i < numCourses; i++) {
    //         if (inDegree[i] == 0) q.offer(i);
    //     }

    //     while (!q.isEmpty()) {
    //         int tmp = q.poll();
    //         cnt += 1;
    //         for (int cn : cnext.get(tmp)) {
    //             inDegree[cn] -= 1;
    //             if (inDegree[cn] == 0) q.offer(cn);
    //         }
    //     }
    //     return cnt == numCourses;
    // }

    // Method 2. DFS - BT
    private Map<Integer, List<Integer>> cprev;
    private int[] visit;

    private boolean dfs(int course) {
        // 3. check for loop or verified results
        if (visit[course] == 1) return false;
        if (visit[course] == 2) return true;
        // 4. update the status & process its preq
        visit[course] = 1;
        for (int cp : cprev.get(course)) {
            if (!dfs(cp)) return false;
        }
        // 5. update to verified & return
        visit[course] = 2;
        return true;
    }

    public boolean canFinish(int numCourses, int[][] preq) {
        this.cprev = new HashMap<>();
        this.visit = new int[numCourses];

        // 1. build preq list for each course
        for (int i = 0; i < numCourses; i++) cprev.put(i, new ArrayList<>());
        for (int[] i : preq) cprev.get(i[0]).add(i[1]);

        // 2. loop - DFS to check each class
        for (int i = 0; i< numCourses; i++) {
            if (!dfs(i)) return false;
        }
        return true;
    }
}